**Python I - Titanic Analysis Group Assignment - Group 4**

---

**Part I: Basic exploratory analysis (3 points)**


---



**Question 1:**
Read the dataset correctly using pandas, using an appropriate column as index. What is the PassengerId of "Mr. Robert William Norman"?

In [4]:
import pandas as pd
df=pd.read_csv('titanic.csv',index_col="PassengerId")

In [2]:
df.head()

Survived  Pclass                                        Name  \
PassengerId                                                                 
1                   0       3                             Mr. Owen Harris   
2                   1       1  Mrs. John Bradley (Florence Briggs Thayer)   
3                   1       3                                 Miss. Laina   
4                   1       1          Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                           Mr. William Henry   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
1              male  22.0      1      0         A/5 21171   7.2500   NaN   
2            female  38.0      1      0          PC 17599  71.2833   C85   
3            female  26.0      0      0  STON/O2. 3101282   7.9250   NaN   
4            female  35.0      1      0            113803  53.1000  C123   
5              male  35.0      0      0            373450   8.0500   NaN   

            Embarked  
PassengerId           
1                  S  
2                  C  
3                  S  
4                  S  
5                  S

In [5]:
# Getting index for row with Passenger name Mr. Robert William Norman since we chose passenger ID as our index 
df.loc[df["Name"] == "Mr. Robert William Norman"].index.tolist()

[235]

---
**Question 2:**
Compute the percentage of survivors out of total number of passengers and the percentage of men and women out of total number of passengers. Paste the answers below.


In [6]:
# survival rate
# rounding method with 2 decimals, 100* to get the percentage
round(100*len(df.loc[(df["Survived"] == 1 )]) / len(df),2)

38.38

In [7]:
# percentage of females
round(100* len(df.loc[(df["Sex"] == "female" )]) / len(df),2)

35.24

In [8]:
# percentage of males
round(100* len(df.loc[(df["Sex"] == "male" )]) / len(df),2)

64.76

---
**Question 3:**
Compute the percentage of survivors by sex (i.e. the percentage of male passengers that survived and female passengers that survived) and the sex distribution of survivors (i.e. percentage of survivors that were men and percentage of survivors that were women). Paste the answers below.


In [9]:
# Simplifying task with function that takes two values: 
# sex refers to what sex's survival rate you want
# population refers to out of what people you want to do that calculation, i.e. male survivors out of all male passengers

def survival_rates(sex,population):
    if population == "survivors":
        return round(100* len(df.loc[(df["Survived"] == 1 )&(df["Sex"] == sex )]) / len(df.loc[(df["Survived"] == 1 )]),2)
    else:
        return round(100* len(df.loc[(df["Survived"] == 1 )&(df["Sex"] == sex )]) / len(df.loc[(df["Sex"] == population )]),2)  

In [10]:
# survival rate of males
survival_rates("male","male")

18.89

In [11]:
# To cross check whether the formula works for male survivors out of all males: 
round(100* len(df.loc[(df["Survived"] == 1 )&(df["Sex"] == "male" )]) / len(df.loc[(df["Sex"] == "male" )]),2)

18.89

In [12]:
# survival rate of females
survival_rates("female","female")

74.2

In [13]:
# percentage of male survivors out of all survivors
survival_rates("male","survivors")

31.87

In [14]:
# To cross check whether the formula works for all male survivors out of all survivors 
round(100* len(df.loc[(df["Survived"] == 1 )&(df["Sex"] == "male" )]) / len(df.loc[(df["Survived"] == 1 )]),2)

31.87

In [15]:
# percentage of female survivors out of all survivors
survival_rates("female","survivors")

68.13

---
**Question 4:**
Display in a 2 x 2 DataFrame the probability of being male/female and surviving/not surviving.

One possible option:
 





           | Survived          | Not survived          |
    Male   | Male & Survived   | Male & Not survived   |
    Female | Female & Survived | Female & Not survived |




Notice that the sum of all values in the table above should be 1 (or 100 %).


In [16]:
# Creating a dummy column with more informative naming for "Survived" variable ("Survived" & "Not Survived")
df["Survived_label"] = df["Survived"]
df.loc[df["Survived"]==0,"Survived_label"] = "Not Survived"
df.loc[df["Survived"]==1,"Survived_label"] = "Survived"

In [17]:
# Pivot table with percentage of survival out of all passengers, by gender
survival_gender = df.pivot_table(
     index="Sex",
     columns="Survived_label",
     values="Survived",
     aggfunc=lambda x: round(len(x)/len(df["Survived"]),2)
     )
survival_gender

Survived_label  Not Survived  Survived
Sex                                   
female                  0.09      0.26
male                    0.53      0.12

In [18]:
# Brining the survivors table into right format
survival_gender.loc[["male","female"],["Survived","Not Survived"]]

Survived_label  Survived  Not Survived
Sex                                   
male                0.12          0.53
female              0.26          0.09

In [19]:
# Exporting survivors table to string
survival_gender.loc[["male","female"],["Survived","Not Survived"]].to_string()

'Survived_label  Survived  Not Survived\nSex                                   \nmale                0.12          0.53\nfemale              0.26          0.09'

---
**Question 5:**
Display in a DataFrame the probability of survival/not survival of all combinations of sex and class.
 
One possible option:

           | | Survived          | Not survived  |
    Male   |1| Male & Survived & Class 1   | ... |
           |2| Male & Survived & Class 2   | ... |
           |3| Male & Survived & Class 3   | ... |
    Female |1| ...                         | ... |
           |2| ...                         | ... |
           |3| ...                         | ... |

Notice that the sum of all values in the table above should be 1 (or 100 %).



In [ ]:
# Pivot table with percentage of survival out of all passengers, by gender and pclass
survival_gender_class = df.pivot_table(
     index=["Sex","Pclass"],
     columns="Survived_label",
     values="Survived",
     aggfunc=lambda x: round(len(x)/len(df["Survived"]),2)
     )
survival_gender_class

Survived_label  Not Survived  Survived
Sex    Pclass                         
female 1                0.00      0.10
       2                0.01      0.08
       3                0.08      0.08
male   1                0.09      0.05
       2                0.10      0.02
       3                0.34      0.05

In [ ]:
# Brining the survivors table into right format
survival_gender_class.loc[["male","female"],["Survived","Not Survived"]]

Survived_label  Survived  Not Survived
Sex    Pclass                         
male   1            0.05          0.09
       2            0.02          0.10
       3            0.05          0.34
female 1            0.10          0.00
       2            0.08          0.01
       3            0.08          0.08

In [ ]:
# Exporting survivors table to string
survival_gender_class.loc[["male","female"],["Survived","Not Survived"]].to_string()

'Survived_label  Survived  Not Survived\nSex    Pclass                         \nmale   1            0.05          0.09\n       2            0.02          0.10\n       3            0.05          0.34\nfemale 1            0.10          0.00\n       2            0.08          0.01\n       3            0.08          0.08'

---
**Part II: Insights and imputation (5 points)**


---




Open-ended question. Exploration of several methods encouraged. More complex and innovative approaches get a higher grade.
You can enrich your analysis with pandas methods and techniques not seen in class, but plots or using libraries other than pandas will not be taken into account for the grade.


**Question 6:** 
(3 Points)

Present three insights about the dataset, each of them relating at least 3 different variables, and support them by code and numbers. Possible examples:

    "Men aged less than 18 were more/less likely to survive than the average passenger" (Sex, Age, Survival)
    "Women with no siblings or spouses paid on average a cheaper/pricier ticket than the average woman" (Sex, SibSp, Fare)
    "Men with a title other than Mr. were more/less likely to have a known (i.e. non-missing) cabin than the average man" (Sex, Name, Cabin)
 
(Using these exact examples is valid, but awards fewer points than proposing original insights)

---

**Insight 1 - Hypothesis:**
(Fare, Embark, Sex) Women overall paid more for their fare, and will pay different fares per embarking location.

In [ ]:
# Pivot table with fare, embark and gender to check fare price differences across groups
df.pivot_table(
    index = ["Sex","Embarked"],
    values=["Fare"],
   aggfunc=["max","min","mean","median","count"])

max     min       mean    median count
                     Fare    Fare       Fare      Fare  Fare
Sex    Embarked                                             
female C         512.3292  7.2250  75.169805  56.92920    73
       Q          90.0000  6.7500  12.634958   7.76875    36
       S         263.0000  7.2500  38.740929  24.15000   203
male   C         512.3292  4.0125  48.262109  24.00000    95
       Q          90.0000  6.7500  13.838922   7.75000    41
       S         263.0000  0.0000  21.711996  10.50000   441

**Insight 1 - Analysis Results:**

*   From the table above, we can see that the majority of Female population embarked from port "S", about 203 women.
*   We can also confirm our hypothesis that depending on the embarking location, you pay more or less for the ticket, with port "C" being the highest cost and port "Q" the lowest in cost on average per ticket.
*   Women pay on average more than men, with a slight exception being port "Q" for the mean, but not for the median.
*   This may be because men apparently have access to cheaper tickets. This could either be because men have special roles / are crew / part of the army.
*   Women would be most likely purely commercial passengers, having access only to the relatively more expensive tickets.

---
**Insight 2 - Hypothesis:**
(Sex, Class, Cabin) Women from higher classes were more likely to have a cabin.

In [ ]:
# Create a column to distinguish whether the passengers has a cabin or not
df["has_Cabin"] = df["Cabin"].notnull()
df.head()

Survived  Pclass                                        Name  \
PassengerId                                                                 
1                   0       3                             Mr. Owen Harris   
2                   1       1  Mrs. John Bradley (Florence Briggs Thayer)   
3                   1       3                                 Miss. Laina   
4                   1       1          Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                           Mr. William Henry   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
1              male  22.0      1      0         A/5 21171   7.2500   NaN   
2            female  38.0      1      0          PC 17599  71.2833   C85   
3            female  26.0      0      0  STON/O2. 3101282   7.9250   NaN   
4            female  35.0      1      0            113803  53.1000  C123   
5              male  35.0      0      0            373450   8.0500   NaN   

            Embarked Survived_label  has_Cabin  
PassengerId                                     
1                  S   Not Survived      False  
2                  C       Survived       True  
3                  S       Survived      False  
4                  S       Survived       True  
5                  S   Not Survived      False

In [ ]:
# Pivot table with gender, pclass and has_Cabin variables to check number of cabins for each sub-group and the ratios between cabin / no-cabin
df.pivot_table(
    index=["Sex","Pclass"],
    columns="has_Cabin",
    values="Name",
    aggfunc="count",
    margins=True
)

has_Cabin      False  True  All
Sex    Pclass                  
female 1          13    81   94
       2          66    10   76
       3         138     6  144
male   1          27    95  122
       2         102     6  108
       3         341     6  347
All              687   204  891

In [ ]:
# Pivot table with percentages of female and male per pclass 
# to verify that class 3 males are a much larger group relative to all males compared to females
df.pivot_table(
    index=["Sex","Pclass"],
    values="Name",
    aggfunc=lambda x: round(len(x)/len(df["Name"]),2),
    margins=True
)

Name
Sex    Pclass      
female 1       0.11
       2       0.09
       3       0.16
male   1       0.14
       2       0.12
       3       0.39
All            1.00

**Insight 2 - Analysis Results:**

*   We can confirm that the higher the Pclass, the less likely it is that the passenger has no cabin. Out of the female passengers from class one, 13 do not have a cabin. For the 144 female passengers from class 3, only 6 have a cabin. So the ratio of "cabin" to "no cabin" significantly differs between classes.
*   The same trend holds true for the male passengers. However, the difference between classes here is much higher than for women. 
*   This effect comes especially from the class 3 men, which make up the majority of the male passengers. 
*   The reasoning is in line with our first finding, that women are more likely to be "commercial passengers", both paying more money for their ticket and likely being from a higher societal class even when purchasing a Pclass 3 ticket. Men from Pclass 3 are most likely more often less wealthy workers or even military & ship crew who will less likely get their own cabin.


---
**Insight 3 - Hypothesis:**
(Parch, SibSp, Survival) People who had more parents/children or siblings/spouses on board were more likely to survive.

In [ ]:
# We first look at people with parents and children on board and the number of survivers among them.
df.pivot_table(
    index=["Parch"],
    columns=["Survived_label"],
    values="Name",
    aggfunc="count",
    margins=True)

Survived_label  Not Survived  Survived  All
Parch                                      
0                      445.0     233.0  678
1                       53.0      65.0  118
2                       40.0      40.0   80
3                        2.0       3.0    5
4                        4.0       NaN    4
5                        4.0       1.0    5
6                        1.0       NaN    1
All                    549.0     342.0  891

**Insight 3 - Analysis Results (Part 1):**

*   We can confirm that people who neither had a parent or child were less likely to survive than people with one or two children / parents.
*   Out of all 678 people without parents or children on board, only 233, so around a third survived. For people with one or two children/parents, the likelyhood to survive was much higher: 50% for those with 2 children/parents and even more than 50% for people with 1 child/parent on board.
*   However, as the number of children on board increases (we can assume 3 means children as one cannot usually have more than 2 parents), the likelihood to survive decreases. Parents with more than 3 children on board were leass likely to survive (some even with 0% chance).
*   This is most likely due to the fact that parents will try to save all of their children which unfortunately decreases their own likelihood of survival.


In [ ]:
# Second, we check people with spouses or siblings on board and the number of survivors among them.
df.pivot_table(
    index=["SibSp"],
    columns=["Survived_label"],
    values="Name",
    aggfunc="count",
    margins=True)

Survived_label  Not Survived  Survived  All
SibSp                                      
0                      398.0     210.0  608
1                       97.0     112.0  209
2                       15.0      13.0   28
3                       12.0       4.0   16
4                       15.0       3.0   18
5                        5.0       NaN    5
8                        7.0       NaN    7
All                    549.0     342.0  891

**Insight 3 - Analysis Results (Part 2):**

*   We see a similar result when looking at people with spouses or siblings on board. People without a spouse or sibling were less likely to survive than people with one sibling or spouse: Roughly 30% without spouse/sibling survived vs. over 50% for people with one spouse/sibling.
*   Again, as the number of siblings increased (again, people will not have more than one spouse), the likelihood of survival diminishes quickly. For 2 siblings, its still roughly 50%, but goes down to 25% percent for 3 siblings. People with 5 or 8 siblings on board had 0% chance of surviving.
*   This can be explained by the same reasoning as when looking at parents/children. People would want to try and save all their siblings, thus reducing the likelihood of their own and the whole group's survival.



---
**Question 7**
(2 Points)

Devise two distinct strategies to impute the missing values in the Age column.
 
(You cannot use machine learning algorithms for the feature engineering part, only manual techniques. This rules out KNN, Deep Learning, sklearn.preprocessing.Imputer, fancyimpute, missingno, and similar.)
 
Some basic ideas (that must not be reused), in increasing order of sophistication:
 
    Fill all the missing values with random values (very bad!)
    Fill all the missing values with the average age of the whole dataset (bad!)
    Fill all the missing values with the median age (so so)
    Fill all the missing values for men with the median age of men, and the missing values of women with the median age of women (better)
 
The objective is to find groupings that yield more granular information about the age.
 
Extra points awarded for proper justifications of the choices made in terms of some quantitative metric (for example, the standard deviation of the age for each sub-group).

In [ ]:
# Getting overview of number of missing values
len(df.loc[df["Age"].isnull()])

177

In [ ]:
# Our starting point is gender as grouping variable. We verify, that we have no missing values in the gender column.
df["Sex"].unique()

array(['male', 'female'], dtype=object)

In order to evaluate the suitability of the sub-group for predicting the age, we need to understand the baseline standard deviation for age based on just gender first. 
Any subgroup we create should ideally have a **smaller standard deviation** (~ <14) than if we just group by gender in order to be a better predictor.

In [ ]:
# Relevant descriptive statistics for gender to understand our baseline
df.pivot_table(
    index=["Sex"],
    values="Age",
    aggfunc=["mean","median","std"])

mean median        std
              Age    Age        Age
Sex                                
female  27.915709   27.0  14.110146
male    30.726645   29.0  14.678201

---
**Strategies & Approach**

Our hypothesis is that the following additional variables will help better predict the age of the passengers in addition to gender:


*   **Title**: The title can indicate whether someone is a child ("Master") or unmarried (usually younger than married) 
*   **Pclass**: The passenger class on the ship is highly correlated with the fare price and serves as a categorical proxy for the societal level that a passenger is in. It may help predict age as there may be differences in the median age between different societal levels
*   **Survived**: Whether a person survived or not may give additional insight into how old a person was (although only in combination with gender, Title and Pclass)



We will use these variables to iteratively improve our prediction power for the missing age values.

---
**Creating first variable "Title" and checking relevance:**

In [ ]:
# First, we verify, that the names do not contain null values, because we will be extracting the titles from the names
df["Name"].isnull().value_counts()

False    891
Name: Name, dtype: int64

In [ ]:
# Create the new variable using the first part of the name before the point (.)
df["Title"] = df["Name"].str.split(".").str[0]
df.head()

Survived  Pclass                                        Name  \
PassengerId                                                                 
1                   0       3                             Mr. Owen Harris   
2                   1       1  Mrs. John Bradley (Florence Briggs Thayer)   
3                   1       3                                 Miss. Laina   
4                   1       1          Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                           Mr. William Henry   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
1              male  22.0      1      0         A/5 21171   7.2500   NaN   
2            female  38.0      1      0          PC 17599  71.2833   C85   
3            female  26.0      0      0  STON/O2. 3101282   7.9250   NaN   
4            female  35.0      1      0            113803  53.1000  C123   
5              male  35.0      0      0            373450   8.0500   NaN   

            Embarked Survived_label  has_Cabin Title  
PassengerId                                           
1                  S   Not Survived      False    Mr  
2                  C       Survived       True   Mrs  
3                  S       Survived      False  Miss  
4                  S       Survived       True   Mrs  
5                  S   Not Survived      False    Mr

In [ ]:
# We make sure that the titles have been correctly extracted and confirm, that all entries in column "Title" are indeed titles
df["Title"].value_counts()

Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Mlle              2
Major             2
Col               2
the Countess      1
Capt              1
Ms                1
Sir               1
Lady              1
Mme               1
Don               1
Jonkheer          1
Name: Title, dtype: int64

**Example 1: Masters**

We verify, that for all "Masters", the standard deviation of age is indeed much smaller than for just gender. Note that Masters can only be male, so we did not need to filter for gender here as it would give us the same result.

In [ ]:
# Standard deviation for age for people with title "Master"
df.loc[df["Title"]=="Master"]["Age"].describe()["std"]

3.6198716433439615

---
**Confirming first variable "Title":**
* By adding the "Title" variable to the grouping condition, we can effectively reduce the standard deviation of age for all subgroups
* Because standard deviation cannot be computed if a group only has one value, we don't see all titles in this table
* We will check whether the missing values apply to other titles than these later on

In [ ]:
# Standard deviation for groups by gender and title to compare with baseline stdv of around 14 for just gender
df.pivot_table(
    index=["Sex","Title"],
    values="Age",
    aggfunc=["std"])

std
                     Age
Sex    Title            
female Miss    12.990292
       Mlle     0.000000
       Mrs     11.433628
male   Col      2.828427
       Dr      12.876335
       Major    4.949747
       Master   3.619872
       Mr      12.708793
       Rev     13.136463

---
**Confirming second variable "Pclass":**
* At this point, we add another variable to the grouping conditions, namely "Pclass"
* We hypothesize that there is an age difference between the different ticket classes which are a proxy for societal classes
* Indeed, we see that for some sub-groups of title and class, the standard deviation of age can be further reduced

In [ ]:
# Standard deviation for groups by gender, title and pclass to compare with standard deviation with just gender & title
df.pivot_table(
    index=["Sex","Title"],
    columns=["Pclass"],
    values="Age",
    aggfunc=["std"])

std                      
Pclass                 1          2          3
Sex    Title                                  
female Miss    12.852308  13.374708   9.697315
       Mlle     0.000000        NaN        NaN
       Mrs     12.643754  10.318525  10.031579
male   Col      2.828427        NaN        NaN
       Dr       9.165151  21.920310        NaN
       Major    4.949747        NaN        NaN
       Master   5.165475   2.342634   3.593608
       Mr      14.148275  11.850977  10.490946
       Rev           NaN  13.136463        NaN

**Cross-check with "fare" - could it be useful to include fare as well?**
* It might be interesting to additionally use fare as a predictor for societal level and thus improve our predictive power for age
* But looking at the fares within the different Pclasses, we see a (non-statistically speaking) significant difference between the fares
* Since we would need to create a categorical variable for fare in order to use it for grouping, we decide to use Pclass as a reasonable proxy instead
* With Pclass, we can most likely capture most of the relevant correlation between societal level and age that could be derived from "fare"

In [ ]:
# Checking correlation between pclass and fare
df.pivot_table(
    index=["Pclass"],
    values="Fare",
    aggfunc=["max","min","mean","median","std"])

max  min       mean   median        std
            Fare Fare       Fare     Fare       Fare
Pclass                                              
1       512.3292  0.0  84.154687  60.2875  78.380373
2        73.5000  0.0  20.662183  14.2500  13.417399
3        69.5500  0.0  13.675550   8.0500  11.778142

---
***Not* onfirming third variable "Survived":**
* Finally, we add "Survived" as a differentiator to our existing groups to check whether it improves the prediction power for age
* Unfortunately, the variable does not provide additional insight into the age gap as it either results in similar or higher standard deviation for most groups
* This may be due to the fact that there is no clear relationship between surviving and gender, title or pclass in that combination
* Including "Survived" would therefore in this variable combination diminish our predictive power for age. We chose to not include it in the final set


In [ ]:
# Checking standard deviation for all sub-groups gender, title, pclass and survived to see whether "survived" improves standard deviation
df.pivot_table(
    index=["Sex","Title"],
    columns=["Pclass","Survived"],
    values="Age",
    aggfunc=["std"])

std                                              \
Pclass                 1                     2                     3   
Survived               0          1          0          1          0   
Sex    Title                                                           
female Miss    33.941125  12.034189   9.899495  13.505097   9.987736   
       Mlle          NaN   0.000000        NaN        NaN        NaN   
       Mrs           NaN  12.519530  14.843629   9.853126   8.480080   
male   Dr            NaN  12.727922  21.920310        NaN        NaN   
       Master        NaN   5.165475        NaN   2.342634   3.480558   
       Mr      14.642111  12.351334  11.681542  14.445299  10.900370   
       Rev           NaN        NaN  13.136463        NaN        NaN   

                          
Pclass                    
Survived               1  
Sex    Title              
female Miss     9.116061  
       Mlle          NaN  
       Mrs     11.081946  
male   Dr            NaN  
       Master   3.989450  
       Mr       7.079075  
       Rev           NaN

---
**Final Sub-Group Variables: Gender, Title and Pclass (Part 1): Mean and Median Comparison**
* Here, we compute the mean and median for the sub-groups created from the final "prediction" variables: gender, title and pclass
* We see that the mean and median are in most cases very close together or even exactly the same
* It is most important to check this for the sub-groups that still posess a rather larger standard deviation for age, namely for any "Mrs", "Mr" and "Rev"
* Although there is only a small difference between mean and median, we choose to use the median to impute age after all

In [ ]:
# Age mean and median by gender, title and pclass to check difference between mean and median
df.pivot_table( 
    index=["Sex","Title"],
    columns=["Pclass"],
    values="Age",
    aggfunc=["mean","median"])

mean                       median            
Pclass                       1          2          3      1     2     3
Sex    Title                                                           
female Dr            49.000000        NaN        NaN   49.0   NaN   NaN
       Lady          48.000000        NaN        NaN   48.0   NaN   NaN
       Miss          30.000000  22.390625  16.123188   30.0  24.0  18.0
       Mlle          24.000000        NaN        NaN   24.0   NaN   NaN
       Mme           24.000000        NaN        NaN   24.0   NaN   NaN
       Mrs           40.882353  33.682927  33.515152   41.5  32.0  31.0
       Ms                  NaN  28.000000        NaN    NaN  28.0   NaN
       the Countess  33.000000        NaN        NaN   33.0   NaN   NaN
male   Capt          70.000000        NaN        NaN   70.0   NaN   NaN
       Col           58.000000        NaN        NaN   58.0   NaN   NaN
       Don           40.000000        NaN        NaN   40.0   NaN   NaN
       Dr            42.000000  38.500000        NaN   44.0  38.5   NaN
       Jonkheer      38.000000        NaN        NaN   38.0   NaN   NaN
       Major         48.500000        NaN        NaN   48.5   NaN   NaN
       Master         5.306667   2.258889   5.350833    4.0   1.0   4.0
       Mr            41.580460  32.768293  28.724891   40.0  31.0  26.0
       Rev                 NaN  43.166667        NaN    NaN  46.5   NaN
       Sir           49.000000        NaN        NaN   49.0   NaN   NaN

---
**Final Sub-Group Variables: Gender, Title and Pclass (Part 2): Checking Sub-Groups for Missing Age Values**

In [ ]:
# Finally, we check which sub-groups are actually missing age values.
df.loc[df["Age"].isnull()].pivot_table(
    index=["Sex","Title"],
    columns=["Pclass"],
    values="Name",
    aggfunc=["count"])

count           
Pclass            1    2     3
Sex    Title                  
female Miss     1.0  2.0  33.0
       Mrs      8.0  NaN   9.0
male   Dr       1.0  NaN   NaN
       Master   NaN  NaN   4.0
       Mr      20.0  9.0  90.0

In [ ]:
# Checking number of records for relevant missing gender, title and pclass sub-groups (Miss, Mrs, Dr, Master, Mr)
prediction_data_count = df.pivot_table(
    index=["Sex","Title"],
    columns=["Pclass"],
    values="Age",
    aggfunc=["count"])
prediction_data_count.loc[[("female","Miss"),("female","Mrs"),("male","Dr"),("male","Master"),("male","Mr")]]

count             
Pclass            1     2      3
Sex    Title                    
female Miss    45.0  32.0   69.0
       Mrs     34.0  41.0   33.0
male   Dr       3.0   2.0    NaN
       Master   3.0   9.0   24.0
       Mr      87.0  82.0  229.0

* We verify that we have enough data to predict the missing ages for the given sub-groups
* Indeed, for all sub-groups, we have existing passenger data to infer an age for the missing values (e.g. 45 people with title "Miss" and Pclass of "1")

---
**Final Sub-Group Variables: Gender, Title and Pclass (Part 3): Creating Final Age Imputation Table**

* Finally, we create the imputation table which we would use to impute the missing values
* We see that this creates a significantly different picture than if we had just used the median ages of 27 (female) and 30 (male) to impute the missing values
* For example, young masters will be much younger and there is a difference in age for the "Missuses" and "Mrs" depending on the Pclass they belong in
* In general, adults and children seem to have significantly different ages - we shall test this theory in the next part

In [ ]:
# Median table for relevant missing gender, title and pclass sub-groups (Miss, Mrs, Dr, Master, Mr)
imputation_table = df.pivot_table(
    index=["Sex","Title"],
    columns=["Pclass"],
    values="Age",
    aggfunc=["median"])
imputation_table.loc[[("female","Miss"),("female","Mrs"),("male","Dr"),("male","Master"),("male","Mr")]]

median            
Pclass             1     2     3
Sex    Title                    
female Miss     30.0  24.0  18.0
       Mrs      41.5  32.0  31.0
male   Dr       44.0  38.5   NaN
       Master    4.0   1.0   4.0
       Mr       40.0  31.0  26.0

---
**Additional Cross-Check: Adult vs. Child Theory**

In [ ]:
# Creating another variable for being an adult or a child 
df["Is_Adult"] = df["Age"]>18
df.head()

Survived  Pclass                                        Name  \
PassengerId                                                                 
1                   0       3                             Mr. Owen Harris   
2                   1       1  Mrs. John Bradley (Florence Briggs Thayer)   
3                   1       3                                 Miss. Laina   
4                   1       1          Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                           Mr. William Henry   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
1              male  22.0      1      0         A/5 21171   7.2500   NaN   
2            female  38.0      1      0          PC 17599  71.2833   C85   
3            female  26.0      0      0  STON/O2. 3101282   7.9250   NaN   
4            female  35.0      1      0            113803  53.1000  C123   
5              male  35.0      0      0            373450   8.0500   NaN   

            Embarked Survived_label  has_Cabin Title  Is_Adult  
PassengerId                                                     
1                  S   Not Survived      False    Mr      True  
2                  C       Survived       True   Mrs      True  
3                  S       Survived      False  Miss      True  
4                  S       Survived       True   Mrs      True  
5                  S   Not Survived      False    Mr      True

In [ ]:
# median and standard deviation for gender and "is_adult" variables to check whether "is_adult" decreases standard deviation
df.pivot_table(
    index=["Sex"],
    columns=["Is_Adult"],
    values="Age",
    aggfunc=["median","std"])

median             std           
Is_Adult  False  True     False       True
Sex                                       
female     13.5  31.0  6.273428  10.834893
male       11.0  31.5  6.684164  12.540078

* We verify that indeed, the median age differs heavily by this categorization
* The standard deviation is also much lower (than initial 14) within both the "kids" and "adults" group, confirming that it is important to make the distinction between kids and adults when imputing age values

In [ ]:
# Combining "is_adult" variable with our other variables to confirm that standard deviation is improved when introducing the "is_adult" variable
df.pivot_table(
    index=["Sex","Title"],
    columns=["Is_Adult","Pclass"],
    values="Age",
    aggfunc=["std"])

std                                                     
Is_Adult          False                           True                      
Pclass                1         2         3          1          2          3
Sex    Title                                                                
female Miss    5.180665  6.252179  6.291580  11.431806   8.732627   5.745313
       Mlle         NaN       NaN       NaN   0.000000        NaN        NaN
       Mrs     0.577350       NaN  1.732051  10.724106   9.950873   8.933020
male   Col          NaN       NaN       NaN   2.828427        NaN        NaN
       Dr           NaN       NaN       NaN   9.165151  21.920310        NaN
       Major        NaN       NaN       NaN   4.949747        NaN        NaN
       Master  5.165475  2.342634  3.593608        NaN        NaN        NaN
       Mr      0.707107  1.032796  1.601282  13.813129  11.447554  10.077039
       Rev          NaN       NaN       NaN        NaN  13.136463        NaN

* We further verify that this also holds true when we add the "Adult" variable to our sub-groups
* We see that the title is a good indicator for the age, given that masters can only be kids, "Cols", "Drs", "Majors" and "Revs" can only be adults
* Pclass further helps making the imputation more precise where we have no clear distinction between adults and kids just from the title